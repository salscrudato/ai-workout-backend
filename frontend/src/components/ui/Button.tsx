import React, { forwardRef, ButtonHTMLAttributes, useMemo } from 'react';
import { motion } from 'framer-motion';
import { clsx } from 'clsx';
// Simple animation variants
const buttonVariants = {
  initial: { scale: 1 },
  hover: { scale: 1.02 },
  tap: { scale: 0.98 },
};

const advancedButtonVariants = {
  initial: { scale: 1, opacity: 1 },
  hover: { scale: 1.02, opacity: 0.9 },
  tap: { scale: 0.98 },
};

export type ButtonVariant = 'primary' | 'secondary' | 'ghost' | 'outline' | 'gradient' | 'premium' | 'luxury' | 'minimal' | 'electric' | 'glass';
export type ButtonSize = 'sm' | 'md' | 'lg' | 'xl';

export interface ButtonProps extends ButtonHTMLAttributes<HTMLButtonElement> {
  variant?: ButtonVariant;
  size?: ButtonSize;
  loading?: boolean;
  loadingText?: string;
  leftIcon?: React.ReactNode;
  rightIcon?: React.ReactNode;
  fullWidth?: boolean;
  children: React.ReactNode;
  animate?: string; // For backward compatibility
  enhancedAnimation?: boolean; // Enable advanced micro-interactions
  ripple?: boolean; // Enable ripple effect
}

const Button = forwardRef<HTMLButtonElement, ButtonProps>(({
  variant = 'primary',
  size = 'md',
  loading = false,
  loadingText,
  leftIcon,
  rightIcon,
  fullWidth = false,
  className,
  children,
  disabled,
  enhancedAnimation = false,
  ripple = false,
  ...props
}, ref) => {
  // Memoize variant classes for better performance
  const variantClasses = useMemo(() => {
    const variants = {
      primary: 'btn-primary micro-bounce focus-premium',
      secondary: 'btn-secondary micro-bounce focus-premium',
      ghost: 'btn-ghost gentle-glow focus-premium',
      outline: 'btn-outline micro-bounce focus-premium',
      gradient: 'gradient-aurora text-white premium-shadow-lg micro-bounce focus-premium',
      premium: 'glass-premium text-primary-700 premium-shadow-lg micro-bounce focus-premium',
      luxury: 'glass-tinted text-primary-800 premium-shadow-xl gentle-glow focus-premium',
      minimal: 'btn-ghost gentle-glow focus-premium',
      electric: 'gradient-ocean text-white premium-shadow-lg micro-bounce focus-premium',
      glass: 'glass-premium text-primary-700 premium-shadow focus-premium'
    };
    return variants[variant] || variants.primary;
  }, [variant]);

  const getSizeClasses = () => {
    const sizes = {
      sm: 'btn-sm',
      md: 'btn-md',
      lg: 'btn-lg',
      xl: 'btn-xl'
    };
    return sizes[size];
  };

  const buttonClasses = clsx(
    'btn',
    variantClasses,
    getSizeClasses(),
    {
      'w-full': fullWidth,
      'opacity-50 cursor-not-allowed': disabled || loading,
    },
    // Enhanced mobile interactions
    'touch-manipulation select-none',
    'active:scale-[0.98] active:transition-transform active:duration-75',
    enhancedAnimation && 'transform-gpu will-change-transform',
    className
  );

  const renderIcon = (icon: React.ReactNode, position: 'left' | 'right') => {
    if (!icon) return null;

    return (
      <span className={clsx(
        'inline-flex items-center justify-center',
        position === 'left' ? 'mr-2' : 'ml-2',
        size === 'sm' ? 'w-4 h-4' : size === 'lg' ? 'w-6 h-6' : 'w-5 h-5'
      )}>
        {icon}
      </span>
    );
  };

  // Choose animation variant based on props
  const animationVariant = enhancedAnimation ? advancedButtonVariants : buttonVariants;

  if (enhancedAnimation) {
    return (
      <motion.button
        ref={ref}
        className={buttonClasses}
        disabled={disabled || loading}
        variants={animationVariant}
        initial="initial"
        whileHover="hover"
        whileTap="tap"
        whileFocus="focus"
        {...(props as any)}
      >
        <div className="flex items-center justify-center relative overflow-hidden">
          {loading ? (
            <motion.div
              className="w-4 h-4 border-2 border-current border-t-transparent rounded-full mr-2"
              animate={{ rotate: 360 }}
              transition={{ duration: 1, repeat: Infinity, ease: "linear" }}
            />
          ) : (
            renderIcon(leftIcon, 'left')
          )}
          <span>{loading && loadingText ? loadingText : children}</span>
          {!loading && renderIcon(rightIcon, 'right')}

          {/* Ripple effect */}
          {ripple && (
            <motion.div
              className="absolute inset-0 bg-white/20 rounded-full scale-0"
              whileTap={{ scale: 4, opacity: [0.5, 0] }}
              transition={{ duration: 0.3 }}
            />
          )}
        </div>
      </motion.button>
    );
  }

  return (
    <button
      ref={ref}
      className={buttonClasses}
      disabled={disabled || loading}
      {...props}
    >
      <div className="flex items-center justify-center">
        {loading ? (
          <div className="w-4 h-4 border-2 border-current border-t-transparent rounded-full animate-spin mr-2" />
        ) : (
          renderIcon(leftIcon, 'left')
        )}
        <span>{loading && loadingText ? loadingText : children}</span>
        {!loading && renderIcon(rightIcon, 'right')}
      </div>
    </button>
  );
});

Button.displayName = 'Button';

export default Button;