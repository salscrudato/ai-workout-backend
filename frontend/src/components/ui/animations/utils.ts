/**
 * Animation utilities and helper functions
 */

// Easing functions
export const easing = {
  // Material Design easing curves
  standard: [0.4, 0.0, 0.2, 1],
  decelerate: [0.0, 0.0, 0.2, 1],
  accelerate: [0.4, 0.0, 1, 1],
  sharp: [0.4, 0.0, 0.6, 1],
  
  // Custom easing curves
  bounce: [0.68, -0.55, 0.265, 1.55],
  elastic: [0.175, 0.885, 0.32, 1.275],
  smooth: [0.25, 0.46, 0.45, 0.94],
  
  // Apple-style easing
  appleEase: [0.25, 0.1, 0.25, 1],
  appleSpring: [0.5, 0, 0.75, 0],
} as const;

// Duration constants
export const duration = {
  instant: 0,
  fast: 150,
  normal: 300,
  slow: 500,
  slower: 750,
  slowest: 1000,
} as const;

// Spring configurations
export const spring = {
  gentle: { type: 'spring', stiffness: 120, damping: 14 },
  wobbly: { type: 'spring', stiffness: 180, damping: 12 },
  stiff: { type: 'spring', stiffness: 210, damping: 20 },
  slow: { type: 'spring', stiffness: 280, damping: 60 },
  molasses: { type: 'spring', stiffness: 280, damping: 120 },
} as const;

// Utility function to create stagger delays
export const createStaggerDelay = (index: number, baseDelay = 0.1) => {
  return index * baseDelay;
};

// Utility function to create random delays for organic feel
export const createRandomDelay = (min = 0, max = 0.5) => {
  return Math.random() * (max - min) + min;
};

// Utility function to interpolate between values
export const interpolate = (
  value: number,
  inputRange: [number, number],
  outputRange: [number, number],
  clamp = true
) => {
  const [inputMin, inputMax] = inputRange;
  const [outputMin, outputMax] = outputRange;
  
  let result = outputMin + ((value - inputMin) / (inputMax - inputMin)) * (outputMax - outputMin);
  
  if (clamp) {
    result = Math.min(Math.max(result, Math.min(outputMin, outputMax)), Math.max(outputMin, outputMax));
  }
  
  return result;
};

// Utility function to create responsive animation values
export const createResponsiveValue = <T>(
  mobile: T,
  tablet: T,
  desktop: T
): T => {
  if (typeof window === 'undefined') return mobile;
  
  const width = window.innerWidth;
  if (width < 768) return mobile;
  if (width < 1024) return tablet;
  return desktop;
};

// Utility function to check if user prefers reduced motion
export const prefersReducedMotion = (): boolean => {
  if (typeof window === 'undefined') return false;
  return window.matchMedia('(prefers-reduced-motion: reduce)').matches;
};

// Utility function to create accessible animations
export const createAccessibleAnimation = <T extends Record<string, any>>(
  animation: T,
  reducedMotionFallback?: Partial<T>
): T => {
  if (prefersReducedMotion()) {
    return {
      ...animation,
      ...reducedMotionFallback,
      transition: {
        ...animation.transition,
        duration: 0.01, // Nearly instant for reduced motion
      },
    };
  }
  return animation;
};

// Utility function to calculate optimal animation duration based on distance
export const calculateDuration = (
  distance: number,
  velocity = 1000 // pixels per second
): number => {
  const baseDuration = distance / velocity;
  return Math.max(0.15, Math.min(baseDuration, 0.8)); // Clamp between 150ms and 800ms
};

// Utility function to create path animations for SVG
export const createPathAnimation = (pathLength: number) => ({
  initial: {
    pathLength: 0,
    opacity: 0,
  },
  animate: {
    pathLength: 1,
    opacity: 1,
    transition: {
      pathLength: { duration: 2, ease: 'easeInOut' },
      opacity: { duration: 0.3 },
    },
  },
});

// Utility function to create morphing animations
export const createMorphAnimation = (
  fromPath: string,
  toPath: string,
  duration = 0.5
) => ({
  initial: { d: fromPath },
  animate: { d: toPath },
  transition: { duration, ease: easing.smooth },
});

// Utility function to create loading skeleton animations
export const createSkeletonAnimation = () => ({
  animate: {
    backgroundPosition: ['0% 50%', '100% 50%', '0% 50%'],
  },
  transition: {
    duration: 2,
    repeat: Infinity,
    ease: 'linear',
  },
});

// Utility function to create floating animations
export const createFloatingAnimation = (
  amplitude = 10,
  duration = 3
) => ({
  animate: {
    y: [-amplitude, amplitude, -amplitude],
  },
  transition: {
    duration,
    repeat: Infinity,
    ease: 'easeInOut',
  },
});

// Utility function to create pulsing animations
export const createPulseAnimation = (
  scale = 1.05,
  duration = 2
) => ({
  animate: {
    scale: [1, scale, 1],
    opacity: [0.7, 1, 0.7],
  },
  transition: {
    duration,
    repeat: Infinity,
    ease: 'easeInOut',
  },
});

// Utility function to create typing cursor animation
export const createCursorAnimation = () => ({
  animate: {
    opacity: [1, 0, 1],
  },
  transition: {
    duration: 1,
    repeat: Infinity,
    ease: 'linear',
  },
});

// Utility function to create page transition variants
export const createPageTransition = (direction: 'left' | 'right' | 'up' | 'down' = 'right') => {
  const directions = {
    left: { x: -100 },
    right: { x: 100 },
    up: { y: -100 },
    down: { y: 100 },
  };

  return {
    initial: {
      opacity: 0,
      ...directions[direction],
    },
    animate: {
      opacity: 1,
      x: 0,
      y: 0,
      transition: {
        duration: 0.4,
        ease: easing.smooth,
      },
    },
    exit: {
      opacity: 0,
      ...directions[direction === 'left' ? 'right' : direction === 'right' ? 'left' : direction === 'up' ? 'down' : 'up'],
      transition: {
        duration: 0.3,
        ease: easing.sharp,
      },
    },
  };
};

// Utility function to create notification animations
export const createNotificationAnimation = (position: 'top' | 'bottom' = 'top') => ({
  initial: {
    opacity: 0,
    y: position === 'top' ? -50 : 50,
    scale: 0.9,
  },
  animate: {
    opacity: 1,
    y: 0,
    scale: 1,
    transition: {
      type: 'spring',
      stiffness: 500,
      damping: 30,
    },
  },
  exit: {
    opacity: 0,
    y: position === 'top' ? -50 : 50,
    scale: 0.9,
    transition: {
      duration: 0.2,
    },
  },
});

// Performance optimization utilities
export const optimizeForPerformance = {
  // Use transform instead of changing layout properties
  useTransform: true,
  // Enable hardware acceleration
  style: {
    transformStyle: 'preserve-3d' as const,
    backfaceVisibility: 'hidden' as const,
  },
  // Reduce motion for better performance on low-end devices
  reduceMotionOnLowEnd: () => {
    if (typeof navigator !== 'undefined') {
      const connection = (navigator as any).connection;
      return connection && (connection.effectiveType === 'slow-2g' || connection.effectiveType === '2g');
    }
    return false;
  },
};
