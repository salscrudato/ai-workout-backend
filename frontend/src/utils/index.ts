/**
 * Consolidated Utilities Index
 * 
 * Central export point for all utility functions and classes.
 * Organized by category for better discoverability and maintainability.
 */

// =============================================================================
// PERFORMANCE UTILITIES
// =============================================================================
export {
  // Core utilities
  debounce,
  throttle,
  rafThrottle,
  
  // Performance monitoring
  performanceMonitor,
  usePerformanceMonitor,
  usePerformanceTracking,
  measureAsync,
  measureSync,
  
  // Optimization utilities
  createIntersectionObserver,
  preloadImage,
  preloadImages,
  optimizeScroll,
  preventLayoutShift,
  enableGPUAcceleration,
  shouldComponentUpdate,
  
  // Device detection
  prefersReducedMotion,
  prefersHighContrast,
  getDevicePixelRatio,
  isTouchDevice,
  getMemoryUsage,
  getNetworkInfo,
  
  // Types
  type PerformanceMetrics,
  type WebVitals,
} from './performanceUtils';

// =============================================================================
// BROWSER COMPATIBILITY
// =============================================================================
export {
  initializeBrowserCompatibility,
  featureDetection,
  browserDetection,
  polyfills,
  fallbacks,
} from './browserCompatibility';

// =============================================================================
// CACHING UTILITIES
// =============================================================================
export {
  apiCache,
  userCache,
  workoutCache,
  imageCache,
  generateCacheKey,
  invalidateCache,
} from './cache';

// =============================================================================
// ASSET LOADING
// =============================================================================
export {
  assetLoader,
  useAssetPreloader,
} from './assetLoader';

// =============================================================================
// ACCESSIBILITY UTILITIES
// =============================================================================
export {
  announceToScreenReader,
  trapFocus,
  restoreFocus,
  getAccessibleName,
  isElementVisible,
  getFocusableElements,
  createAriaLiveRegion,
  updateAriaLiveRegion,
  removeAriaLiveRegion,
  handleKeyboardNavigation,
  createSkipLink,
  enhanceFormAccessibility,
  checkColorContrast,
  validateAccessibility,
} from './accessibility';

// =============================================================================
// USER INTERACTION
// =============================================================================
export {
  hasUserInteraction,
  onUserInteraction,
  enableVibration,
  enableAudio,
  enableNotifications,
  trackInteraction,
  getInteractionMetrics,
} from './userInteraction';

// =============================================================================
// PROFILE UTILITIES
// =============================================================================
export {
  validateProfile,
  calculateProfileCompleteness,
  getProfileRecommendations,
  formatProfileData,
  sanitizeProfileInput,
} from './profileUtils';

// =============================================================================
// LOGGING
// =============================================================================
export {
  logger,
  createLogger,
  LogLevel,
  type LogEntry,
  type LoggerConfig,
} from './logger';

// =============================================================================
// COMMON UTILITY FUNCTIONS
// =============================================================================

/**
 * Format file size in human readable format
 */
export const formatFileSize = (bytes: number): string => {
  if (bytes === 0) return '0 Bytes';
  
  const k = 1024;
  const sizes = ['Bytes', 'KB', 'MB', 'GB'];
  const i = Math.floor(Math.log(bytes) / Math.log(k));
  
  return parseFloat((bytes / Math.pow(k, i)).toFixed(2)) + ' ' + sizes[i];
};

/**
 * Format duration in human readable format
 */
export const formatDuration = (ms: number): string => {
  if (ms < 1000) return `${Math.round(ms)}ms`;
  if (ms < 60000) return `${(ms / 1000).toFixed(1)}s`;
  if (ms < 3600000) return `${Math.floor(ms / 60000)}m ${Math.floor((ms % 60000) / 1000)}s`;
  
  const hours = Math.floor(ms / 3600000);
  const minutes = Math.floor((ms % 3600000) / 60000);
  return `${hours}h ${minutes}m`;
};

/**
 * Generate a random ID
 */
export const generateId = (prefix = 'id'): string => {
  return `${prefix}_${Math.random().toString(36).substr(2, 9)}`;
};

/**
 * Deep clone an object
 */
export const deepClone = <T>(obj: T): T => {
  if (obj === null || typeof obj !== 'object') return obj;
  if (obj instanceof Date) return new Date(obj.getTime()) as unknown as T;
  if (obj instanceof Array) return obj.map(item => deepClone(item)) as unknown as T;
  if (typeof obj === 'object') {
    const clonedObj = {} as T;
    for (const key in obj) {
      if (obj.hasOwnProperty(key)) {
        clonedObj[key] = deepClone(obj[key]);
      }
    }
    return clonedObj;
  }
  return obj;
};

/**
 * Check if value is empty (null, undefined, empty string, empty array, empty object)
 */
export const isEmpty = (value: any): boolean => {
  if (value == null) return true;
  if (typeof value === 'string') return value.trim().length === 0;
  if (Array.isArray(value)) return value.length === 0;
  if (typeof value === 'object') return Object.keys(value).length === 0;
  return false;
};

/**
 * Capitalize first letter of a string
 */
export const capitalize = (str: string): string => {
  return str.charAt(0).toUpperCase() + str.slice(1).toLowerCase();
};

/**
 * Convert camelCase to kebab-case
 */
export const camelToKebab = (str: string): string => {
  return str.replace(/([a-z0-9]|(?=[A-Z]))([A-Z])/g, '$1-$2').toLowerCase();
};

/**
 * Convert kebab-case to camelCase
 */
export const kebabToCamel = (str: string): string => {
  return str.replace(/-([a-z])/g, (g) => g[1].toUpperCase());
};

/**
 * Clamp a number between min and max values
 */
export const clamp = (value: number, min: number, max: number): number => {
  return Math.min(Math.max(value, min), max);
};

/**
 * Linear interpolation between two values
 */
export const lerp = (start: number, end: number, factor: number): string => {
  return (start + (end - start) * factor).toString();
};

/**
 * Map a value from one range to another
 */
export const mapRange = (
  value: number,
  inMin: number,
  inMax: number,
  outMin: number,
  outMax: number
): number => {
  return ((value - inMin) * (outMax - outMin)) / (inMax - inMin) + outMin;
};

/**
 * Check if code is running in development mode
 */
export const isDevelopment = (): boolean => {
  return process.env.NODE_ENV === 'development';
};

/**
 * Check if code is running in production mode
 */
export const isProduction = (): boolean => {
  return process.env.NODE_ENV === 'production';
};

/**
 * Check if code is running in browser
 */
export const isBrowser = (): boolean => {
  return typeof window !== 'undefined';
};

/**
 * Check if code is running on server
 */
export const isServer = (): boolean => {
  return typeof window === 'undefined';
};

// =============================================================================
// CONSTANTS
// =============================================================================

export const BREAKPOINTS = {
  xs: 475,
  sm: 640,
  md: 768,
  lg: 1024,
  xl: 1280,
  '2xl': 1536,
} as const;

export const ANIMATION_DURATIONS = {
  fast: 150,
  normal: 300,
  slow: 500,
} as const;

export const Z_INDEX = {
  dropdown: 1000,
  sticky: 1020,
  fixed: 1030,
  modal: 1040,
  popover: 1050,
  tooltip: 1060,
  toast: 1070,
} as const;

export const CACHE_KEYS = {
  USER_PROFILE: 'user_profile',
  WORKOUT_HISTORY: 'workout_history',
  EQUIPMENT_LIST: 'equipment_list',
  APP_SETTINGS: 'app_settings',
} as const;
