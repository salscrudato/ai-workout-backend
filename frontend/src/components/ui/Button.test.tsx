/**
 * Button Component Tests
 * 
 * Comprehensive test suite for the Button component demonstrating
 * best practices and testing patterns.
 */

import React from 'react';
import { describe, it, expect, vi } from 'vitest';
import { screen, fireEvent, waitFor } from '@testing-library/react';
import { renderWithProviders } from '../../test/utils';
import {
  testComponentRendering,
  testComponentProps,
  testComponentAccessibility,
  testComponentInteractions,
  testComponentPerformance,
  testComponentSnapshots,
} from '../../test/templates';
import Button from './Button';
import type { ButtonProps } from './Button';

// =============================================================================
// BASIC RENDERING TESTS
// =============================================================================

describe('Button Component', () => {
  // Test basic rendering
  testComponentRendering(Button, { children: 'Test Button' });

  // =============================================================================
  // PROPS TESTING
  // =============================================================================

  testComponentProps(Button, [
    {
      prop: 'variant',
      value: 'primary',
      expectedBehavior: 'applies primary styling',
      testFn: (container) => {
        const button = container.querySelector('button');
        expect(button).toHaveClass('bg-gradient-to-r', 'from-blue-600', 'to-blue-700');
      },
    },
    {
      prop: 'variant',
      value: 'secondary',
      expectedBehavior: 'applies secondary styling',
      testFn: (container) => {
        const button = container.querySelector('button');
        expect(button).toHaveClass('bg-neutral-100', 'text-neutral-900');
      },
    },
    {
      prop: 'size',
      value: 'sm',
      expectedBehavior: 'applies small size styling',
      testFn: (container) => {
        const button = container.querySelector('button');
        expect(button).toHaveClass('px-3', 'py-1.5', 'text-sm');
      },
    },
    {
      prop: 'size',
      value: 'lg',
      expectedBehavior: 'applies large size styling',
      testFn: (container) => {
        const button = container.querySelector('button');
        expect(button).toHaveClass('px-6', 'py-3', 'text-lg');
      },
    },
    {
      prop: 'loading',
      value: true,
      expectedBehavior: 'shows loading state',
      testFn: (container) => {
        const button = container.querySelector('button');
        expect(button).toBeDisabled();
        expect(container.querySelector('.animate-spin')).toBeInTheDocument();
      },
    },
    {
      prop: 'disabled',
      value: true,
      expectedBehavior: 'disables the button',
      testFn: (container) => {
        const button = container.querySelector('button');
        expect(button).toBeDisabled();
        expect(button).toHaveClass('opacity-50', 'cursor-not-allowed');
      },
    },
    {
      prop: 'fullWidth',
      value: true,
      expectedBehavior: 'applies full width styling',
      testFn: (container) => {
        const button = container.querySelector('button');
        expect(button).toHaveClass('w-full');
      },
    },
  ]);

  // =============================================================================
  // INTERACTION TESTS
  // =============================================================================

  testComponentInteractions(
    Button,
    [
      {
        name: 'calls onClick when clicked',
        action: async (container) => {
          const button = container.querySelector('button')!;
          fireEvent.click(button);
        },
        expectation: (container) => {
          // This would be tested with a mock function
        },
      },
      {
        name: 'handles keyboard activation',
        action: async (container) => {
          const button = container.querySelector('button')!;
          button.focus();
          fireEvent.keyDown(button, { key: 'Enter' });
        },
        expectation: (container) => {
          // This would be tested with a mock function
        },
      },
    ],
    { children: 'Test Button' }
  );

  // =============================================================================
  // ACCESSIBILITY TESTS
  // =============================================================================

  testComponentAccessibility(Button, { children: 'Test Button' });

  // =============================================================================
  // PERFORMANCE TESTS
  // =============================================================================

  testComponentPerformance(Button, { children: 'Test Button' });

  // =============================================================================
  // SNAPSHOT TESTS
  // =============================================================================

  testComponentSnapshots(Button, [
    { name: 'default', props: { children: 'Default Button' } },
    { name: 'primary variant', props: { variant: 'primary', children: 'Primary Button' } },
    { name: 'secondary variant', props: { variant: 'secondary', children: 'Secondary Button' } },
    { name: 'loading state', props: { loading: true, children: 'Loading Button' } },
    { name: 'disabled state', props: { disabled: true, children: 'Disabled Button' } },
    { name: 'with icons', props: { leftIcon: '←', rightIcon: '→', children: 'Icon Button' } },
  ]);

  // =============================================================================
  // SPECIFIC BUTTON TESTS
  // =============================================================================

  describe('Button Specific Behavior', () => {
    it('renders children correctly', () => {
      renderWithProviders(<Button>Test Content</Button>);
      expect(screen.getByText('Test Content')).toBeInTheDocument();
    });

    it('handles click events', () => {
      const handleClick = vi.fn();
      renderWithProviders(<Button onClick={handleClick}>Click Me</Button>);
      
      const button = screen.getByRole('button');
      fireEvent.click(button);
      
      expect(handleClick).toHaveBeenCalledTimes(1);
    });

    it('prevents click when disabled', () => {
      const handleClick = vi.fn();
      renderWithProviders(
        <Button onClick={handleClick} disabled>
          Disabled Button
        </Button>
      );
      
      const button = screen.getByRole('button');
      fireEvent.click(button);
      
      expect(handleClick).not.toHaveBeenCalled();
    });

    it('prevents click when loading', () => {
      const handleClick = vi.fn();
      renderWithProviders(
        <Button onClick={handleClick} loading>
          Loading Button
        </Button>
      );
      
      const button = screen.getByRole('button');
      fireEvent.click(button);
      
      expect(handleClick).not.toHaveBeenCalled();
    });

    it('shows loading text when provided', () => {
      renderWithProviders(
        <Button loading loadingText="Please wait...">
          Submit
        </Button>
      );
      
      expect(screen.getByText('Please wait...')).toBeInTheDocument();
      expect(screen.queryByText('Submit')).not.toBeInTheDocument();
    });

    it('renders left icon', () => {
      renderWithProviders(
        <Button leftIcon={<span data-testid="left-icon">←</span>}>
          Button with Left Icon
        </Button>
      );
      
      expect(screen.getByTestId('left-icon')).toBeInTheDocument();
    });

    it('renders right icon', () => {
      renderWithProviders(
        <Button rightIcon={<span data-testid="right-icon">→</span>}>
          Button with Right Icon
        </Button>
      );
      
      expect(screen.getByTestId('right-icon')).toBeInTheDocument();
    });

    it('supports custom className', () => {
      renderWithProviders(
        <Button className="custom-class">Custom Button</Button>
      );
      
      const button = screen.getByRole('button');
      expect(button).toHaveClass('custom-class');
    });

    it('forwards ref correctly', () => {
      const ref = React.createRef<HTMLButtonElement>();
      renderWithProviders(<Button ref={ref}>Ref Button</Button>);
      
      expect(ref.current).toBeInstanceOf(HTMLButtonElement);
    });

    it('supports all button HTML attributes', () => {
      renderWithProviders(
        <Button
          type="submit"
          form="test-form"
          data-testid="custom-button"
          aria-label="Custom button"
        >
          HTML Attributes Button
        </Button>
      );
      
      const button = screen.getByTestId('custom-button');
      expect(button).toHaveAttribute('type', 'submit');
      expect(button).toHaveAttribute('form', 'test-form');
      expect(button).toHaveAttribute('aria-label', 'Custom button');
    });
  });

  // =============================================================================
  // ANIMATION TESTS
  // =============================================================================

  describe('Animation Behavior', () => {
    it('applies hover animations when enhancedAnimation is enabled', () => {
      renderWithProviders(
        <Button enhancedAnimation>Animated Button</Button>
      );
      
      const button = screen.getByRole('button');
      expect(button).toHaveClass('transition-all');
    });

    it('applies ripple effect when ripple is enabled', async () => {
      renderWithProviders(
        <Button ripple>Ripple Button</Button>
      );
      
      const button = screen.getByRole('button');
      fireEvent.click(button);
      
      // Wait for ripple animation to start
      await waitFor(() => {
        const ripple = button.querySelector('.ripple-effect');
        expect(ripple).toBeInTheDocument();
      });
    });
  });

  // =============================================================================
  // VARIANT TESTS
  // =============================================================================

  describe('Button Variants', () => {
    const variants: Array<ButtonProps['variant']> = [
      'primary',
      'secondary',
      'ghost',
      'outline',
      'gradient',
      'premium',
      'luxury',
      'minimal',
      'electric',
      'glass',
    ];

    variants.forEach((variant) => {
      it(`renders ${variant} variant correctly`, () => {
        renderWithProviders(
          <Button variant={variant}>{variant} Button</Button>
        );
        
        const button = screen.getByRole('button');
        expect(button).toBeInTheDocument();
        
        // Each variant should have specific classes
        // This would be more specific based on actual implementation
        expect(button.className).toBeTruthy();
      });
    });
  });

  // =============================================================================
  // SIZE TESTS
  // =============================================================================

  describe('Button Sizes', () => {
    const sizes: Array<ButtonProps['size']> = ['sm', 'md', 'lg', 'xl'];

    sizes.forEach((size) => {
      it(`renders ${size} size correctly`, () => {
        renderWithProviders(
          <Button size={size}>{size} Button</Button>
        );
        
        const button = screen.getByRole('button');
        expect(button).toBeInTheDocument();
        
        // Each size should have specific padding and text size classes
        expect(button.className).toBeTruthy();
      });
    });
  });
});
