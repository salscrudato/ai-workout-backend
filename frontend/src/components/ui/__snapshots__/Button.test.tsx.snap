// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Button Component > Snapshots > matches snapshot for default 1`] = `
<button
  class="btn btn-primary micro-bounce focus-premium btn-md touch-manipulation select-none active:scale-[0.98] active:transition-transform active:duration-75"
>
  <div
    class="flex items-center justify-center"
  >
    <span>
      Default Button
    </span>
  </div>
</button>
`;

exports[`Button Component > Snapshots > matches snapshot for disabled state 1`] = `
<button
  class="btn btn-primary micro-bounce focus-premium btn-md opacity-50 cursor-not-allowed touch-manipulation select-none active:scale-[0.98] active:transition-transform active:duration-75"
  disabled=""
>
  <div
    class="flex items-center justify-center"
  >
    <span>
      Disabled Button
    </span>
  </div>
</button>
`;

exports[`Button Component > Snapshots > matches snapshot for loading state 1`] = `
<button
  class="btn btn-primary micro-bounce focus-premium btn-md opacity-50 cursor-not-allowed touch-manipulation select-none active:scale-[0.98] active:transition-transform active:duration-75"
  disabled=""
>
  <div
    class="flex items-center justify-center"
  >
    <div
      class="w-4 h-4 border-2 border-current border-t-transparent rounded-full animate-spin mr-2"
    />
    <span>
      Loading Button
    </span>
  </div>
</button>
`;

exports[`Button Component > Snapshots > matches snapshot for primary variant 1`] = `
<button
  class="btn btn-primary micro-bounce focus-premium btn-md touch-manipulation select-none active:scale-[0.98] active:transition-transform active:duration-75"
>
  <div
    class="flex items-center justify-center"
  >
    <span>
      Primary Button
    </span>
  </div>
</button>
`;

exports[`Button Component > Snapshots > matches snapshot for secondary variant 1`] = `
<button
  class="btn btn-secondary micro-bounce focus-premium btn-md touch-manipulation select-none active:scale-[0.98] active:transition-transform active:duration-75"
>
  <div
    class="flex items-center justify-center"
  >
    <span>
      Secondary Button
    </span>
  </div>
</button>
`;

exports[`Button Component > Snapshots > matches snapshot for with icons 1`] = `
<button
  class="btn btn-primary micro-bounce focus-premium btn-md touch-manipulation select-none active:scale-[0.98] active:transition-transform active:duration-75"
>
  <div
    class="flex items-center justify-center"
  >
    <span
      class="inline-flex items-center justify-center mr-2 w-5 h-5"
    >
      ←
    </span>
    <span>
      Icon Button
    </span>
    <span
      class="inline-flex items-center justify-center ml-2 w-5 h-5"
    >
      →
    </span>
  </div>
</button>
`;
